### ISSUE TITLE
Introduce `#[swift_bridge(Hashable)]` attribute for opaque Rust types

### ISSUE THREAD

**Issue by chinedufn** [chinedufn/swift-bridge#77]
Given the following bridge:

```rust
#[swift_bridge::bridge]
mod ffi {
    extern "Rust" {
        #[swift_bridge(Hashable)]
        type SomeType;
    }
}

#[derive(Hash, PartialEq, Eq, PartialOrd, Ord)]
struct SomeType;
```

We should emit some Swift along the lines of:

```swift
extension SomeTypeRef: Hashable {
    public func hash(into hasher: inout Hasher) {
        hasher.combine(__swift_bridge__$SomeType$_hash(rhs.ptr))
    }
}
```

And on the Rust side we'd generate something like:

```rust
#[export_name = "__swift_bridge__$SomeType$_hash"]
pub extern "C" fn __swift_bridge__SomeType__hash (
    this: *const super::SomeType,
) -> u64 {
    let mut s = DefaultHasher::new();
    {unsafe &*this}.hash(&mut s);
    s.finish()
}
```

**Comment by chinedufn** [comment]
Sure!

First, check out the book chapter on contributing new types, just to get a sense of the different parts of the codebase.

https://chinedufn.github.io/swift-bridge/contributing/adding-support-for-a-signature/index.html

---

We can add a `mod hashable` where we can define a bridge module for testing the `Hashable` attribute.

https://github.com/chinedufn/swift-bridge/blob/4c08e05b666f1cddcd194411b995bfbfbec7ab26/crates/swift-integration-tests/src/opaque_type_attributes.rs#L2-L3

Similar to how we test other attributes.

https://github.com/chinedufn/swift-bridge/blob/4c08e05b666f1cddcd194411b995bfbfbec7ab26/crates/swift-integration-tests/src/opaque_type_attributes/copy.rs#L1-L16

---

Then we can add an integration test to this `OpaqueTypeAttributeTests.swift`.

https://github.com/chinedufn/swift-bridge/blob/4831b68bf7a8f4457ebd960a1b7b984a9b8bceb7/SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunnerTests/OpaqueTypeAttributeTests.swift#L36-L43

The new test can verify that the types `Hashable` implementation works.

Here's how we test `[swift_bridge(Identifiable)]`

https://github.com/chinedufn/swift-bridge/blob/607e30579ac4de1fab5ff66838253cf3d07ddfbb/SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunnerTests/FunctionAttributeIdentifiableTests.swift#L24-L36

---

We can add a `hashable: bool` field and parse it in a similar the same way to how we parse the `already_declared: bool`. We can use `#[swift_bridge(Hashable)]`.

https://github.com/chinedufn/swift-bridge/blob/54b9c09337701dd7580010d1d5ef6527b855ee98/crates/swift-bridge-ir/src/parse/parse_extern_mod.rs#L717-L740

https://github.com/chinedufn/swift-bridge/blob/c75d6e744ac64e21145afbe4d251486f861e90cd/crates/swift-bridge-ir/src/parse/parse_extern_mod/opaque_type_attributes.rs#L17-L29

---

We can add a `mod identifiable_codegen_tests` https://github.com/chinedufn/swift-bridge/blob/2bffcb08c1f0ca8bbcba9c78c5ae28a9cddfb632/crates/swift-bridge-ir/src/codegen/codegen_tests.rs#L30-L48

Here's an example of a codegen test https://github.com/chinedufn/swift-bridge/blob/41f93586cf3eb7cbe441e257a464f4136e0ddcbb/crates/swift-bridge-ir/src/codegen/codegen_tests/function_attribute_codegen_tests.rs#L185-L256

---

From there we can get the tests passing by modifying our codegen https://github.com/chinedufn/swift-bridge/blob/93bfb436e5f8df9169866ed8ae8c643175f343a6/crates/swift-bridge-ir/src/codegen.rs#L4-L6 .

---

Please let me know if you have any questions!

**Comment by NiwakaDev** [comment]
@chinedufn, thank you for your polite response.

I came up with a question. In your first comment, You provided an example of a struct SomeTypeRef that conforms to the Hashable protocol. Don't a struct called SomeType need to conforms?

Here’s an example of a struct called SomeType that conforms to the Hashable protocol:

```Swift
struct SomeType {
   let value: Int 
}

extension SomeType: Hashable {
    public func hash(into hasher: inout Hasher) {
       hasher.combine(self.x)
    }
}
```

**Comment by chinedufn** [comment]
Ah, good question.

`SomeType` and `SomeTypeMut` inherit from `SomeTypeRef`, so only `SomeTypeRef` needs to have the `Hashable` implementation.

See https://chinedufn.github.io/swift-bridge/bridge-module/opaque-types/index.html#owned-ref-and-refmut for more information.

**Comment by NiwakaDev** [comment]
Hi, @chinedufn 

While working on this issue, I got the following error. 

```
error: type 'SomeTypeRef' does not conform to protocol 'Equatable'
extension SomeTypeRef: Hashable{
```

According to [Apple Developer Documentation](https://developer.apple.com/documentation/swift/hashable), ```SomeTypeRef``` needs to conform to ```Equatable```. So, I might need to work on #72 first. Can I do this one too?

**Comment by NiwakaDev** [comment]
Thank you for your quick response.

>yeah implementing swift_bridge(Equatable) makes sense.

Sure! I'll do this first.

**Comment by chinedufn** [comment]
Closed by #140
